### PR TITLE
Add connection poisoning to aws-smithy-client

### PR DESCRIPTION
**Pull request by rcoh**
## Motivation and Context

- https://github.com/awslabs/aws-sdk-rust/issues/160

## Description
Built on top of two changes to Hyper, this adds the concept of connection poisoning to Smithy client. To do this, we needed machinery to get access to the "connection metadata" used by the request. This is done in a connector agnostic way with the `CaptureSmithyConnection` bridge.

A separate layer considers the error classification then invokes poison as necessary. To support testing this, `WireLeveMock` was created, a mocking facility that binds to 0.0.0.0.

For AWS services, this behavior is **enabled by default**. For generic-smithy clients, this behavior is **disabled by default**.

## Testing
- integration test in S3
- detailed tests of specific behavior in aws-smithy-client

## Checklist

- [x] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates
- [x] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: rust-runtime/aws-smithy-client/src/hyper_ext.rs [hunk 1]
```diff
@@ -108,7 +117,30 @@ use tower::{BoxError, Service};
 /// see [the module documentation](crate::hyper_ext).
 #[derive(Clone, Debug)]
 #[non_exhaustive]
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-client/src/hyper_ext.rs:119]: Not needed anymore:
```suggestion
```

File: rust-runtime/aws-smithy-types/src/retry.rs [hunk 2]
```diff
@@ -230,6 +259,20 @@ pub struct RetryConfig {
...
+///
+/// By default, when a transient error is encountered, the connection in use will be poisoned. This
+/// behavior can be disabled by setting [`ReconnectMode::NoReconnect`] instead.
+#[derive(Debug, Clone, PartialEq, Copy)]
+pub enum ReconnectMode {
+    /// Reconnect on [`ErrorKind::TransientError`]
+    ReconnectOnTransientError,
+
+    /// Disable reconnects
+    NoReconnect,
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-types/src/retry.rs:275]: The name here is probably good enough, but I could see how it could potentially cause confusion with reconnecting on hang up or something.

**Review by jdisanti — APPROVED** [review]
Nice work on this!

File: CHANGELOG.next.toml [hunk 3]
```diff
@@ -269,3 +269,28 @@ message = "Smithy members named `send` were previously renamed to `send_value` a
...
+message = """Reconnect on transient errors.
+
+Note: **this behavior is disabled by default for generic clients**. It can be enabled with
+`aws_smithy_client::Builder::reconnect_on_transient_errors`
+
+If a transient error (timeout, 500, 503, 503) is encountered, the connection will be evicted from the pool and will not
+be reused.
+"""
+references = ["aws-sdk-rust#160", "smithy-rs#2445"]
+meta = { "breaking" = false, "tada" = false, "bug" = false }
```
Comment (82marbag) [inline CHANGELOG.next.toml:295]: Missing target

File: rust-runtime/aws-smithy-types/src/retry.rs (hunk 2 above)
Comment (82marbag) [reply rust-runtime/aws-smithy-types/src/retry.rs:275]: Maybe `NoResuseBadConnection`? Usage is clearer from the tests / examples

**Review by 82marbag — APPROVED** [review]

**Review by ysaito1001 — APPROVED** [review]
`WireLevelTestConnection` is pretty cool!

File: CHANGELOG.next.toml (hunk 3 above)
Comment (rcoh) [reply CHANGELOG.next.toml:313]: ah—we changed the default to `all` at some point which is why this isn't an error

File: rust-runtime/aws-smithy-client/src/hyper_ext.rs [hunk 4]
```diff
@@ -107,8 +116,30 @@ use tower::{BoxError, Service};
...
-#[non_exhaustive]
-pub struct Adapter<C>(HttpReadTimeout<hyper::Client<ConnectTimeout<C>, SdkBody>>);
+pub struct Adapter<C> {
+    client: HttpReadTimeout<hyper::Client<ConnectTimeout<C>, SdkBody>>,
+}
+
+/// Extract a smithy connection from a hyper CaptureConnection
+fn extract_smithy_connection(capture_conn: &CaptureConnection) -> Option<ConnectionMetadata> {
+    let capture_conn = capture_conn.clone();
+    if let Some(conn) = capture_conn.clone().connection_metadata().as_ref() {
```
Comment (crisidev) [inline rust-runtime/aws-smithy-client/src/hyper_ext.rs:126]: Do you really need the double clone here?

File: rust-runtime/aws-smithy-client/src/hyper_ext.rs [hunk 5]
```diff
@@ -107,8 +116,30 @@ use tower::{BoxError, Service};
...
+        let mut extensions = Extensions::new();
+        conn.get_extras(&mut extensions);
+        let http_info = extensions.get::<HttpInfo>();
+        let smithy_connection = ConnectionMetadata::new(
+            conn.is_proxied(),
+            http_info.map(|info| info.remote_addr()),
+            move || match capture_conn.connection_metadata().as_ref() {
+                Some(conn) => conn.poison(),
+                None => tracing::trace!("no connection existed to poison"),
```
Comment (crisidev) [inline rust-runtime/aws-smithy-client/src/hyper_ext.rs:135]: Maybe rephrase this error message like `no connection found that should be poisoned` ?

File: rust-runtime/aws-smithy-client/src/test_connection.rs [hunk 6]
```diff
@@ -271,6 +271,347 @@ where
...
+                                .lock()
+                                .unwrap()
+                                .push(RecordedEvent::Response(next_event.clone()));
+                            if next_event == ReplayedEvent::Timeout {
+                                tracing::info!("{} is poisoned", remote_addr);
+                                poisoned_conns.lock().unwrap().insert(remote_addr);
+                            }
+                            tracing::debug!("replying with {:?}", next_event);
+                            let event = generate_response_event(next_event).await;
+                            dbg!(event)
```
Comment (crisidev) [inline rust-runtime/aws-smithy-client/src/test_connection.rs:508]: Is this a leftover or do we want to debug every time we spinup a new mock?

File: rust-runtime/aws-smithy-client/src/test_connection.rs [hunk 7]
```diff
@@ -271,6 +271,347 @@ where
...
+
+        fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+            Poll::Ready(Ok(()))
+        }
+
+        fn call(&mut self, req: Name) -> Self::Future {
+            let sock_addr = self.socket_addr;
+            let log = self.log.clone();
+            Box::pin(async move {
+                println!("looking up {:?}, replying with {:?}", req, sock_addr);
```
Comment (crisidev) [inline rust-runtime/aws-smithy-client/src/test_connection.rs:605]: Same as above: is this a leftover?
